ystia/yorc
# Feature/gh 209 kubernetes persistent volume claim support

# Pull Request description

## Description of the change

### What I did
Added support of kubernetes PersistentVolumeClaim 

### How I did it

### How to verify it

To prepare for testing this PR, you can run this little script :
**Make sure you have a yorc configuration file in your home directory (~/config.yorc.yaml) configured for kubernetes.**
```bash
#/bin/bash
a4c_version=2.1.1
git_branch="feature/GH-209-Kubernetes-persistent-volume-claim-support"

#Prepare for tests
mkdir testGH-PR && cd testGH-PR &&\
git clone git@github.com:ystia/yorc.git && git clone git@github.com:ystia/yorc-a4c-plugin.git &&\ 
cd yorc-a4c-plugin && git checkout ${git_branch} && mvn clean install -DskipTests -Dalien4cloud.version=${a4c_version} &&\
cd .. && unzip -j yorc-a4c-plugin/distribution/target/alien4cloud-yorc-plugin-distribution-3.2.0-SNAPSHOT.zip &&\
wget https://fastconnect.org/maven/service/local/repositories/opensource/content/alien4cloud/alien4cloud-dist/${a4c_version}/alien4cloud-dist-${a4c_version}-dist.tar.gz &&\ 
tar -xzf alien4cloud-dist-${a4c_version}-dist.tar.gz &&\
cd yorc &&\
git checkout ${git_branch} &&\
export SKIP_TESTS=1 && make tools && make &&\

#Launch binaries
cd build/bin/ && eval "./consul agent -dev -ui&" && sleep 2 &&\
cd ../.. && eval "./yorc server -c ~/config.yorc.yaml&" &&\
export JAVA_EXT_OPTIONS="-Dhttp.nonProxyHosts=\"127.0.0.1|localhost\""
cd ../alien4cloud && eval "./alien4cloud.sh&"
exit 0
```
#### Once alien is started
1. Upload zip of the plugin generated in testGH-PR folder and enable it
2. Execute this lines to configure k8s location :
```bash
export ALIEN_VERSION=2.1.1
wget https://raw.githubusercontent.com/ystia/docker-images/master/alien4cloud/scripts/utils.bash && chmod +x utils.bash &&\
wget -O - https://raw.githubusercontent.com/ystia/docker-images/master/alien4cloud/scripts/create_orchestrator.bash |bash &&\
wget -O - https://raw.githubusercontent.com/ystia/docker-images/master/alien4cloud/scripts/configure_orchestrator.bash |bash &&\
wget -O - https://raw.githubusercontent.com/ystia/docker-images/master/alien4cloud/scripts/enable_orchestrator.bash |bash &&\
wget -O - https://raw.githubusercontent.com/ystia/docker-images/master/alien4cloud/scripts/create_config_location.bash |bash 
```
3. Add "PersistentVolumeClaimSource"
4. Upload a kubernetes sample type and topology using volume, for example "yorc-a4c-plugin/tosca-samples/org/ystia/yorc/samples/kube/yorc" and "yorc-a4c-plugin/tosca-samples/org/ystia/yorc/samples/kube/topology/simple-yorc"
5. On "Matching" section, make sure that volume is match to PersistentVolumeClaim and the size of the volume is specified
6. Deploy the app  ! 
- [x] Connect to GKE console and make sure that the volume and corresponding volume claim are well created

#### Once all work well
1. Undeploy the app
- [x] Connect to GKE console and make sure that the volume and corresponding volume claim are removed

### Description for the changelog


## Applicable Issues
Issue https://github.com/ystia/yorc/issues/209
